server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -131,16 +148,87 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
         return routing.primary() && routing.active() == false && getInactivePrimaryHealth(routing) == ClusterHealthStatus.YELLOW;
     }
 
+    /**
+     * Diagnose well known error states that we know will never result in a successful allocation.
+     * @return Diagnosis result to be returned with actions for a user to take
+     */
+    private static Diagnosis diagnoseReason(ShardRouting routing, ClusterState state) {
+        return Optional.ofNullable(state.metadata().index(routing.index()))
+            .map(indexMetadata -> {
+                // Check: Data Tier conflicts with required attributes
```

Comment:
Currently we're only checking for data tier preference agreeing with allocation filtering - This is trying to detect a well known allocation snag that users experience but it's not necessarily the _cause_ of a shard being unallocated, just something we know will keep it from being assigned.